src/site/content/en/blog/progressively-enhance-your-pwa/index.md

```diff
@@ -0,0 +1,1051 @@
+    requestWakeLock();
+  }
+};
+
+document.addEventListener('visibilitychange', handleVisibilityChange);
+document.addEventListener('fullscreenchange', handleVisibilityChange);
+```
+
+Yes, this is a progressive enhancement, so I only need to load it when the browser
+supports the API.
+
+```js
+if ('wakeLock' in navigator && 'request' in navigator.wakeLock) {
+  import('./wake_lock.mjs');
+}
+```
+In Fugu Greetings, there's an insomnia checkbox that, when checked, keeps the
+screen awake.
+
+<figure class="w-figure">
+  <img class="w-screenshot"
+       src="10000201000009C4000005DBCD9D10DD0A745315.png"
+       alt="The insomnia checkbox, if checked, keeps the screen awake.">
+  <figcaption class="w-figcaption">
+    The insomnia checkbox keeps app awake.
+  </figcaption>
+</figure>
+
+## The Idle Detection API
+
+At times, even if you stare at the screen for hours,
+it's just useless.
+The [Idle Detection API](https://web.dev/idle-detection/) allows the app to detect user idle time.
+If the user is idle for too long, the app resets to the initial state
+and clears the canvas.
+This API is currently gated behind the
+[notifications permission](https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission),
+since a lot of production use cases of idle detection are notifications-related,
+for example, to only send a notification to a device the user is currently actively using.
+
+After making sure that the notifications permission is granted, I then instantiate the
+idle detector.
+I register an event listener that listens for idle changes, which includes the user and
+the screen state.
+The user can be active or idle,
+and the screen can be unlocked or locked.
+If the user is idle, the canvas clears.
+I give the idle detector a threshold of 60 seconds.
+
+```js
+const idleDetector = new IdleDetector();
+idleDetector.addEventListener('change', () => {
+  const userState = idleDetector.userState;
+  const screenState = idleDetector.screenState;
+  console.log(`Idle change: ${userState}, ${screenState}.`);
+  if (userState === 'idle') {
+    clearCanvas();
+  }
+});
+
+await idleDetector.start({
+  threshold: 60000,
+  signal,
+});
+```
+
+And as always, I only load this code when the browser supports it.
+
+```js
+if ('IdleDetector' in window) {
+  import('./idle_detection.mjs');
+}
+```
+
+In the Fugu Greetings app, the canvas clears when the Ephemeral checkbox is
+checked and the user is idle for for too long.
+
+<figure class="w-figure">
+  <img class="w-screenshot"
+       src="10000201000009C4000005DBB930B1D3F8F5D1DA.png"
+       alt="Fugu Greetings app with a cleared canvas after the user has been idle for too long.">
+  <figcaption class="w-figcaption">
+    When the ephemeral checkbox is checked and the user has been idle for too long, the canvas is cleared.
+  </figcaption>
+</figure>
+
+## Closing
+
+Phew, what a ride. So many APIs in just one sample app.
```

Comment:
A long journey, but informative!

Reply:
Thank you :-) The idea is to show as many APIs in as non-contrived context as possible.